### PR TITLE
 K8SPXC-641 Update proxy users on each node

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-641](https://jira.percona.com/browse/K8SPXC-641) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| haproxy | passed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| scaling-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| smart-update | failed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/5d2945eed8feacec0a02a517c713d61f662b19a8
image: `perconalab/percona-xtradb-cluster-operator:PR-795-5d2945ee`

File: pkg/controller/pxc/users.go [hunk 1]
```diff
@@ -354,54 +355,69 @@ func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBClus
...
-	um, err := users.NewManager(cr.Name+"-proxysql-unready."+cr.Namespace+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
-	if err != nil {
-		return errors.Wrap(err, "new users manager")
-	}
-	defer um.Close()
-
-	err = um.UpdateProxyUsers(proxyUsers)
-	if err != nil {
-		return errors.Wrap(err, "update proxy users")
+		um, err := users.NewManager(pod.Status.PodIP+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
```
Comment (delgod) [inline pkg/controller/pxc/users.go:411]: please lets try in the next way and without fetching Pod
```suggestion
		um, err := users.NewManager(cr.Name + "-proxysql-" + strconv.Itoa(i)
+"."+cr.Name+"-proxysql-unready."+cr.Namespace+":6032", "proxyadmin", string(internalSysSecretObj.Data["proxyadmin"]))
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| scaling | passed |
| recreate | passed |
| haproxy | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| restore-to-encrypted-cluster | passed |
| affinity | passed |
| self-healing-advanced | passed |
| one-pod | passed |
| auto-tuning | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/02cf27c8782fa8f05dd30f0ca3d60c26d02ca220
image: `perconalab/percona-xtradb-cluster-operator:PR-795-02cf27c8`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| big-data | passed |
| storage | passed |
| scaling | passed |
| haproxy | passed |
| recreate | passed |
| init-deploy | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| upgrade-proxysql | passed |
| monitoring-2-0 | passed |
| scaling-proxysql | passed |
| demand-backup | passed |
| security-context | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| smart-update | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| scheduled-backup | failed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/67944b64db596ed9614171f0a4a68d5f1f402102
image: `perconalab/percona-xtradb-cluster-operator:PR-795-67944b64`

File: pkg/controller/pxc/users.go (hunk 1 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/73ab8492c1953ae3cd796e194383109390277233

File: pkg/controller/pxc/users.go [hunk 2]
```diff
@@ -354,54 +355,55 @@ func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBClus
...
-	}
-	if len(errb.Bytes()) > 0 {
-		return errors.New("syncusers: " + errb.String())
+
+	for i := 0; i < int(cr.Spec.PXC.Size); i++ {
+		pod := corev1.Pod{}
+		err := r.client.Get(context.TODO(),
+			types.NamespacedName{
+				Namespace: cr.Namespace,
+				Name:      cr.Name + "-proxysql-" + strconv.Itoa(i),
```
Comment (qJkee) [inline pkg/controller/pxc/users.go:370]: will it work if we have different pxc and proxysql size?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| scaling | passed |
| recreate | passed |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| init-deploy | passed |
| scaling-proxysql | failed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| scheduled-backup | failed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/90798a0bd09df4e2357e2df818bc5b93e9ee7636
image: `perconalab/percona-xtradb-cluster-operator:PR-795-90798a0b`

File: pkg/controller/pxc/users.go [hunk 3]
```diff
@@ -354,54 +355,55 @@ func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBClus
...
-	}
-	var errb, outb bytes.Buffer
-	err = r.clientcmd.Exec(&pod, "proxysql", []string{"proxysql-admin", "--syncusers"}, nil, &outb, &errb, false)
-	if err != nil {
-		return errors.Errorf("exec syncusers: %v / %s / %s", err, outb.String(), errb.String())
-	}
-	if len(errb.Bytes()) > 0 {
-		return errors.New("syncusers: " + errb.String())
+
+	for i := 0; i < int(cr.Spec.PXC.Size); i++ {
```
Comment (delgod) [inline pkg/controller/pxc/users.go:365]:
```suggestion
	for i := 0; i < int(cr.Spec.ProxySQL.Size); i++ {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/pxc/users.go (hunk 3 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go]: done https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a

File: pkg/controller/pxc/users.go (hunk 2 above)
Comment (nonemax) [reply pkg/controller/pxc/users.go:370]: fixed https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| recreate | passed |
| storage | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |
| tls-issue-self | failed |
| scheduled-backup | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a
image: `perconalab/percona-xtradb-cluster-operator:PR-795-6e5e4509`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| scaling | passed |
| big-data | passed |
| operator-self-healing | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a
image: `perconalab/percona-xtradb-cluster-operator:PR-795-6e5e4509`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| big-data | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a
image: `perconalab/percona-xtradb-cluster-operator:PR-795-6e5e4509`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |
| haproxy | passed |
| scaling | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| limits | passed |
| big-data | passed |
| security-context | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| upgrade-proxysql | passed |
| tls-issue-cert-manager | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/795/commits/6e5e4509b1cb4ab8ff070822cc598f8ea3f8e08a
image: `perconalab/percona-xtradb-cluster-operator:PR-795-6e5e4509`

**Review by qJkee — APPROVED** [review]

**Pull request merged**